Clean OSAL API

ismilak opened the pull request:
Cleaning OSAL API and make it more portable for other platforms.
Fixing function parameter signatures and add additional error handling where the OSAL APIs are used.
Removing linux header references, and using only OSAL to make the lib core independent on the platform.

ismilak committed:
Remove POSIX/linux headers from sources

ismilak committed:
Replace int ret type with osal_basetype_t

ismilak committed:
Implement integer status return type for thread create function

ismilak committed:
Fix alarm fired call

ismilak committed:
Check return value of thread creation

ismilak committed:
Change osal_recvfrom return type to osal_ssize_t

ismilak committed:
Remove trailing whitespaces

ismilak committed:
Fix type definitions in 'osal/freertos/osal_common.h'

manojnacsl commented on osal/linux/osal_common.c:
```diff
@@ -64,21 +65,25 @@ osal_task_t osal_task_create (
-   pthread_attr_setstacksize (&attr, PTHREAD_STACK_MIN + stacksize);
+   
+   if (thread == NULL) {
+       return -1;
```
Good to use MACROS for readability and consistency.
```
#define SUCCESS 1
#define FAILURE 0
#define ERROR -1
```

manojnacsl replied:
Yes for OSAL, all such defines can be in the common header file(proposed - osal/common.h) Currently agent code has similar defines in src/csmpagent/csmp.h

manojnacsl commented on osal/linux/osal_common.c:
```diff
@@ -43,17 +43,18 @@ static void osal_alarm_fired();
  *****************************************************************************/
-osal_task_t osal_task_create (
+osal_basetype_t osal_task_create(
+   osal_task_t * thread,
```
Any specific reason for passing *thread as parameter to osal_task_create()

ismilak replied:
Yep, my reason are:
- to avoid returning a deep copy of a structure
- standardise the API return value

manojnacsl replied:
Thanks for clarifying.

manojnacsl commented on src/coap/coapserver.c:
```diff
@@ -90,7 +82,9 @@ int coapserver_listen(uint16_t sport, recv_handler_t recv_handler)
   m_server_opened = true;
-  recvt_id_task = osal_task_create(NULL, 0, 0, recv_thread, NULL);
+  ret = osal_task_create(&recvt_id_task, NULL, 0, 0, recv_thread, NULL);
+  assert(ret == 0);
```
Good to add debug logs for failure scenarios.

manojnacsl commented on src/coap/coapserver.c:
```diff
@@ -90,7 +82,9 @@ int coapserver_listen(uint16_t sport, recv_handler_t recv_handler)
+  ret = osal_task_create(&recvt_id_task, NULL, 0, 0, recv_thread, NULL);
+  assert(ret == 0);
+
   return 0;
```
Return values to be replaced by appropriate macros in all instances.

ismilak replied:
Related to our previous discussion and recommendations

manojnacsl replied:
Yes, all such instances could be replaced with valid defines. Thanks.

manojnacsl reviewed (commented):
Please share the test strategy/process being followed to test FreeRTOS specific changes and also to test existing Linux implementation(just to ensure nothing is broken for Linux)

manojnacsl commented on osal/linux/osal_common.c:
```diff
@@ -64,21 +65,25 @@ osal_task_t osal_task_create (
-   ret = pthread_detach(thread);
+   ret = pthread_detach(*thread);
    if (ret != 0){
        return (-ret);
```
Please add a debug log to indicate the failure.

manojnacsl commented on osal/linux/osal_common.c:
```diff
@@ -64,21 +65,25 @@ osal_task_t osal_task_create (
    }
 
-   return thread;
+   return 0;
```
Please use a MACRO for return code.

manojnacsl commented on sample/CsmpAgentLib_sample.c:
```diff
@@ -16,10 +16,7 @@
 
 #include <stdio.h>
 #include <stdbool.h>
-#include <stdlib.h>
```
For any change being done to the sample App it is recommended to test the sample App build for other target platforms as well to ensure nothing is broken.

manojnacsl commented on src/coap/coapclient.c:
```diff
@@ -70,7 +69,9 @@ int coapclient_open(response_handler_t response_handler)
   m_client_opened = true;
-  recvt_id_task = osal_task_create(NULL, 0, 0, recv_fn, NULL);
+  ret = osal_task_create(&recvt_id_task, NULL, 0, 0, recv_fn, NULL);
+  assert(ret == 0);
```
Please add a debug log for failure.

manojnacsl commented on src/coap/coapserver.c:
```diff
@@ -59,6 +50,7 @@ int coapserver_listen(uint16_t sport, recv_handler_t recv_handler)
 {
   osal_socket_handle_t sockfd;
   osal_sockaddr listen_addr;
+  osal_basetype_t ret = 0;
```
Good to use a MACRO to indicate return codes.

ismilak commented:
> Please share the test strategy/process being followed to test FreeRTOS specific changes and also to test existing Linux implementation(just to ensure nothing is broken for Linux)

Hi Everyone,

It has been tested manually.
- Build test: make and ./build.sh
- Smoke test: I went through step by step on the tutorial doc and the agent was able to connect to and communicate with the FND. I have a Cisco FND VM on a linux host, and I built and executed the agent on Raspberry Pi4 and  on the linux host. (Ubuntu 22.04)

manojnacsl commented:
> > Please share the test strategy/process being followed to test FreeRTOS specific changes and also to test existing Linux implementation(just to ensure nothing is broken for Linux)
> 
> @manojnacsl May I return the question? Do you have test infrastructure/suites that we could rely on?

@silabs-ThibautC - Currently we do not have a separate test suite as such. But the usual test practice we have been following for Linux target is to have the Linux agent connect to FND instance to test - registration, periodic metric refresh, and GET operations of the TLVs.

ismilak committed:
Add defined return values to osal API

ismilak committed:
Add debug prints

ismilak committed:
Refactor osal directory and create separated common content

ismilak commented:
Hi @manojnacsl and @paduffy !

I solved all the findings from the current and previous PR reviews:

- common osal API return code definitions
- debug prints for API return code checking 
- osal source and header file refactor, common header and osal.c/h file hierarchy modified based on @manojnacsl proposal

Could you take a look at this again? 

THX!

ismilak committed:
Add doxy comments for osal_comman.h

silabs-ThibautC commented on osal/osal_common.h:
```diff
@@ -6,62 +6,483 @@
+/****************************************************************************
+ * @fn     osal_fd_isset 
+ *
+ * @brief  modifies  the  contents of the sets
```
It is not what it does.

ismilak replied:
Fixed

silabs-ThibautC commented on osal/osal_common.h:
```diff
@@ -6,62 +6,483 @@
+/****************************************************************************
+ * @fn    osal_print_formatted_ip
+ *
+ * @brief print formatted ip v6 address from sockadd for debugging purpose
```
IPv6

ismilak replied:
Fixed

silabs-ThibautC reviewed (changes requested):
* Run a spellchecker
* Check from double withespaces

ismilak commented on osal/osal_common.h:
```diff
@@ -6,62 +6,483 @@
+/****************************************************************************
+ * @fn        osal_task_cancel
+ *
+ * @brief     function requests that thread be canceled. 
```
It depends on the cancel type (other pthread call) There are asynchronous and deferred cancel type for the pthread_setcanceltype API. These functions are wrappers of pthread functions. Set cancel type is a dummy implementation,  pthread_cancel is a wrapper of vTaskDelete() for FreeRTOS

silabs-ThibautC replied:
I'll create an issue to cover the improvement of the API documentation. It is not the initial topic of that PR and that text is not your doing

ismilak committed:
Clean and fix doxy comments in osal_common.h, removing comments from osal.c

silabs-ThibautC commented on linux.target:
```diff
@@ -16,12 +16,14 @@ LIBOBJ = $(patsubst %.c,%.o,$(LIBOBJ_nodir))
-INC += osal/linux/
+LIBSRC += osal/linux/osal.c
+INC += \
+-I osal/linux/ \
```
Do we really need that directory in the path? It is counterintuitive. Everything in that folder should have a local scope.

ismilak replied:
I think yes, because I cannot use `#include ../osal.h` in osal_common.h since I have to specify the target dir: linux, freertos and later efr32. Relative path cannot be used by this way, because osal.h parent directory can be different based on different platforms

silabs-ThibautC replied:
If we have to do it it means that we are doing something wrong in osal_common.h. The public header should not be platform-dependent. The typedefs must be abstracted in osal_common.h and we should not need to include that platform dependent header.

silabs-ThibautC replied:
Indeed maybe we won't be able to do it here. It will not work cleanly with the thread id definition for instance. Still something bothers me. 

Consider that comment resolved, I'm completing my proposal about the new file architecture.

ismilak commented on osal/osal_common.h:
```diff
@@ -6,62 +6,483 @@
+ * input parameters
+ *  @param[in] thread address to store thread ID
+ *  @param[in] name name of the task
+ *  @param[in] priotiy of task 
```
Added and for stack size explanation also

ismilak commented on osal/osal_common.h:
```diff
@@ -6,62 +6,483 @@
+ * input parameters
+ *  @param[in] thread address to store thread ID
+ *  @param[in] name name of the task
+ *  @param[in] priotiy of task 
```
Done

ismilak commented on osal/osal_common.h:
```diff
@@ -6,62 +6,483 @@
+ *
+ * @brief Allows a program to monitor multiple file descriptors, waiting until
+ *        one or more of the file descriptors become "ready" for some class of 
+ *        I/O operatio 
```
Fixed

ismilak commented on osal/osal_common.h:
```diff
@@ -6,62 +6,483 @@
 osal_basetype_t osal_inet_pton(int af, const char *src, void *dst);
+
+/****************************************************************************
+ * @fn      osal_select
```
The select API is not used in select contexts, since works with only one file descriptor in coapclient.c and coapserver.c in the library sources. It can be entirely replaced with blocking recvfrom(), because there isn't any other "readable" descriptor in the implementation.

ismilak commented on osal/osal_common.h:
```diff
@@ -6,62 +6,483 @@
+ *                     they are ready for reading 
+ *  @param[in] writefds The file descriptors in this set are watched to see 
+ *                      if they are ready for writing.
+ *  @param[in] exceptfds The  file descriptors in this set are watched for 
```
As we discussed already, select is used but I think not necessarily.
https://github.com/CiscoDevNet/csmp-agent-lib/pull/13#discussion_r1599551182

ismilak commented:
> * Run a spellchecker
> * Check from double withespaces

I executed spellchecker, and I standardised the comments, removed unnecessary white spaces etc.
I removed the doxy comments from osal.c, I note that I just copied comments from it, that was the original location of code documentation. If it's required, I can copy back the checked and fixed doxy comments into the source file, but I think we should avoid the duplications. It makes the c file cleaner also.

ismilak committed:
Replace int symbols with base type and socket type in osal

ismilak committed:
Replace file descriptor context with socket descriptor

ismilak committed:
Rename symbols to standardise osal API names

ismilak committed:
Removing unused calloc and realloc API

silabs-ThibautC commented:
@ismilak @manojnacsl 
I apologize because I'm re-opening a topic that I think you already closed but I'm really not a big fan of the file organization you agreed upon. 
* there's a file name duplication (osal.c, osal.h) which can be confusing. Today their are only two. A pair would be added for each new platform supported
* osal.h looks like a public API but it is actually a platform-dependent header.
* osal_common.h does not look like a public API header but it actually is.

osal/
├── osal.h
├── freertos/
 |             └ public API header, can be included anywhere else in the project
│   ├── osal_freertos.c
│   └── osal_freertos.h 
│            └ I'm not sure we need that file see my comment above. It should only be included in osal_freertos.c
├── linux/
│   ├── osal_linux.c
│   └── osal_linux.h
└── silabs/
    ├── osal_silabs.c
    └── osal_silabs.h
    
 If we agree that we could merge osal_platform.h into osal_platform.c it could even be simplified:
osal/
├── osal.h
├── osal_freertos.c
├── osal_linux.c
├── osal_silabs.c

ismilak commented:
> @ismilak @manojnacsl I apologize because I'm re-opening a topic that I think you already closed but I'm really not a big fan of the file organization you agreed upon.
> 
> * there's a file name duplication (osal.c, osal.h) which can be confusing. Today their are only two. A pair would be added for each new platform supported
>   osal.h looks like a public API but it is actually a platform-dependent header.
> * osal_common.h does not look like a public API header but it actually is.
> 
> **Edit: Agreed that we need to keep the platform-specific header. Forget my two previous proposals.**
> 
> osal/ ├── osal.h ├── freertos/ │ ├── osal_freertos.c │ └── osal_platform_types.h │ └ Only contains type definitions, could be renamed to make it clear. ├── linux/ │ ├── osal_linux.c │ └── osal_platform_types.h └── silabs/ ├── osal_silabs.c └── osal_platform_types.h
> 
> Invalid proposals
> 
> osal/ ├── osal.h ├── freertos/ | └ public API header, can be included anywhere else in the project │ ├── osal_freertos.c │ └── osal_freertos.h │ └ I'm not sure we need that file see my comment above. It should only be included in osal_freertos.c ├── linux/ │ ├── osal_linux.c │ └── osal_linux.h └── silabs/ ├── osal_silabs.c └── osal_silabs.h
> 
> If we agree that we could merge osal_platform.h into osal_platform.c it could even be simplified: osal/ ├── osal.h ├── osal_freertos.c ├── osal_linux.c ├── osal_silabs.c

I did the same proposal (except file names) ~1.5 month ago, what was dropped then Guys: https://github.com/CiscoDevNet/csmp-agent-lib/pull/12#issuecomment-2032051462

I don't really understand why do we need to include the platform names in the osal_xxxxx.c file names, because it's specified in the directory name, at this point in the make file you can specialise the linking for the particular platform, and disagree with the osal_platform_types.h file name too, why do we need to include "_platform_ " in it? 

Back to my proposal, I think we should use the simplest possible file names and "architecture". I note that @manojnacsl recommendation was the same as mine except the file names, so finally it seems we defined the same architecture in a long story :)

silabs-ThibautC commented:
`why do we need to include "platform " in it?`
Because it only contains types that are platform-specific. All the other types should remain in osal.h

`why do we need to include the platform names` `the simplest possible file names`
Using the same filename multiple times is error-prone, I think we can all agree on that. 

Like said, osal_common.h does not look like a public header. It looks like a header common to osal C files that should not be included elsewhere. I'd naturally include osal.h, not osal_common.h in my project.

It does not contain anything but type definitions. See it as the equivalent of stdint.h.

ismilak commented:
> `why do we need to include "platform " in it?` Because it only contains type definitions that are platform-specific. All the other types should remain in osal.h
> 
> `why do we need to include the platform names` `the simplest possible file names` Using the same filename multiple times is error-prone, I think we can all agree on that.
> 
> Like said, osal_common.h does not look like a public header. It looks like a header common to osal C files that should not be included elsewhere. I'd naturally include osal.h, not osal_common.h in my project.
> 
> It does not contain anything but type definitions. See it as the equivalent of stdint.h.

`Because it only contains type definitions that are platform-specific. All the other types should remain in osal.h` -> Entire osal is platform specific, since it includes network programming interface, OS API and utils, like timer, etc. (IMHO). I would say, "osal_types.h" is crystal clear for anybody, see my 1.5 months old proposal ...

`Using the same filename multiple times is error-prone, I think we can all agree on that. ` -> just one of the several counter-examples is port.c in FreeRTOS. I note that if you try to build an efr32 project with linux version of osal.c, you will see the error at the build time immediately, so I don't think  it's a critical error-prone implementation as it is now.

 `See it as the equivalent of stdint.h.` As I mentioned, 1.5 months ago I recommended the same solution: osal_types.h (https://github.com/CiscoDevNet/csmp-agent-lib/pull/12#issuecomment-2032051462)
 
 Just to clarify:
 - I agree with having a types.h, the name of this header is not a top-prio topic I think (again, I have proposed the same earlier)
 - I agree with using osal.h name instead of osal_common.h in the library sources as the common header
 - I disagree with _osal_platform_types.h_ name, I think _osal_types.h_ it's quite enough
 - I disagree with osal_xxxx.c source file names for xxxx platforms, since the same file name is arranged to platform specific directories already, so I think it's overhead, just we can use osal.c instead (my example is FreeRTOS port.c)

silabs-ThibautC commented:
`Entire osal is platform specific` 
That is true today, it might not be always the case. For instance there's a scenario where we move all the RTOS functions in a single C file osal_cmsis.c for instance. That file could be re-used by several platforms. 
And of course, osal.h is not.

That said, as long as we rename osal_common.h and osal.h, respectively osal.h and osal_(something-or-nothing_)types.h I'm fine.

silabs-ThibautC commented on osal/linux/osal.h:
```diff
@@ -0,0 +1,59 @@
+
+#ifndef __OSAL_H
+#define __OSAL_H
+#include <stdarg.h>
```
We probably have too many headers included here. There are more files included than types defined. 

I understand stdint.h signal.h, semaphore.h, select.h, pthread.h, socket.h and netinit/in.h
I think ssize_t is included from stdio.h but that remains to be checked

I think all the others could be removed or moved to the C file.

silabs-ThibautC commented on osal/freertos/osal.h:
```diff
@@ -0,0 +1,59 @@
+
+#ifndef __OSAL_H
+#define __OSAL_H
+#include <stdarg.h>
```
Same as https://github.com/CiscoDevNet/csmp-agent-lib/pull/13/files#r1601548384 ?

silabs-ThibautC commented:
As discussed in the meeting today (05/15):
osal/
├── osal.h
├── freertos/
│ ├── osal_freertos.c
│ └── osal_platform_types.h
│ └ Only contains type definitions, could be renamed to make it clear.
├── linux/
│ ├── osal_linux.c
│ └── osal_platform_types.h
└── silabs/
├── osal_silabs.c
└── osal_platform_types.h

ismilak committed:
Renaming osal artifacts

ismilak commented:
Test with Cisco FND was successful:
- No build issues, warnings on Ubuntu 22.04
- Sample app connected to the FND, registration request was succesful
- It sends the metrics to the FND

manojnacsl commented:
> Hi @manojnacsl and @paduffy !
> 
> I solved all the findings from the current and previous PR reviews:
> 
>     * common osal API return code definitions
> 
>     * debug prints for API return code checking
> 
>     * osal source and header file refactor, common header and osal.c/h file hierarchy modified based on @manojnacsl proposal
> 
> 
> Could you take a look at this again?
> 
> THX!

@ismilak The latest code pushed looks good and inline to our syncup discussion. Approving.
